### PR TITLE
target/openrisc: Implement proper exception vector calculation

### PR DESCRIPTION
**Pull request by mithro**
* Add support for `EVBAR` as a CPU feature.
 * Add support for `SR[EPH]`.

### CODE REVIEW

File: target/openrisc/interrupt.c [hunk 1]
```diff
@@ -57,7 +57,9 @@ void openrisc_cpu_do_interrupt(CPUState *cs)
     env->tlb->cpu_openrisc_map_address_code = &cpu_openrisc_get_phys_nommu;
 
     if (cs->exception_index > 0 && cs->exception_index < EXCP_NR) {
-        env->pc = (cs->exception_index << 8);
+        hwaddr vect_pc = (cs->exception_index << 8) | env->evbar;
+        printf("Exception 0x%x - vectoring to 0x%x\n", cs->exception_index, (unsigned int)vect_pc);
```
Comment (stffrdhrn) [inline target/openrisc/interrupt.c]: I think we can remove the printf?

**Review by stffrdhrn — APPROVED** [review]
Looks good. Ill merge if you want to remove the printf, good, otherwise I will before sending upstream.

File: target/openrisc/interrupt.c (hunk 1 above)
Comment (mithro) [reply target/openrisc/interrupt.c]: Done.

**Comment by mithro** [comment]
Hi @stffrdhrn ,

I've updated the pull request with the change you requested. I've also added support for the EPH bit and made EVBAR support dependent on the configuration. 

I'm wondering however if the EVBAR support should be something which a the `xxx_initf` call `set_feature` to enable? It is unclear to me why `openrisc_cpu_reset` sets `cpu->env.cpucfgr` directly rather than using the value of `cpu->feature`?

Tim 'mithro' Ansell

File: target/openrisc/cpu.c [hunk 2]
```diff
@@ -51,7 +51,7 @@ static void openrisc_cpu_reset(CPUState *s)
     s->exception_index = -1;
 
     cpu->env.upr = UPR_UP | UPR_DMP | UPR_IMP | UPR_PICP | UPR_TTP;
-    cpu->env.cpucfgr = CPUCFGR_OB32S | CPUCFGR_OF32S;
+    cpu->env.cpucfgr = cpu->feature;
```
Comment (stffrdhrn) [inline target/openrisc/cpu.c]: thanks I was thinking about this change too, it was strange to have the mask set 2 times.

If you are doing this should we remove the CPUCFGR_OV64S variables too?

**Comment by stffrdhrn** [comment]
This looks good, one more suggestion that will help with the upstream qemu mailing list review.   They usually dont like small fixes to previous patches.  Could we squash these ?

```
target-openrisc: Implement EVBAR register          \_ squash
target-openrisc: Make EVBAR conditional on CPUCFGR /
target-openrisc: Implement EPH bit
target-openrisc: On reset set CPUCFGR to features    - remove CPUCFGR?
target-openrisc: Make EVBAR a OpenRisc feature.
```

`Make EVBAR conditional on CPUCFGR`  This looks like something that was missed from the original change? could we squash them?

The `On reset set CPUCFGR to features` is a small change on its own and could be thought as part of the `Make EVBAR a OpenRisc feature` change.  But if we also remote call `CPUCFGR` defines then it makes sense on its own. i.e. it could be called `Consolidate CPUCFGR and features`

**Comment by mithro** [comment]
It looks like you have squashed and pushed these changes into your branch? Is there anything more you need from me to get this upstream?

**Comment by stffrdhrn** [comment]
Hello, i haven't done it yet. But i was messing around with some other changes on my branch which might be causing this conflict sorry about that. If you agree with those changes i can just so them?

**Comment by mithro** [comment]
@stffrdhrn I've updated this pull request. I've also moved the CPUCFGR change into a separate branch to send. 

Going to run the checking patch tools on them now.

**Comment by mithro** [comment]
@stffrdhrn I think this patch is now ready to send to the mailing list, what do you think? If you are happy with it, I'll send it out tomorrow morning.

**Comment by stffrdhrn** [comment]
I think this looks good. One thing I mentioned in the irc room was since we are adding evbar to env, we need to serialize it in machine.c as well.

**Pull request closed without merging**